New Lesson - Automating with Openconfig

jnpr-raylam opened the pull request:
It's ready for review. Thanks.

Stage1: Install Openconfig Yang module on Junos box
Stage2: Provision Openconfig using CLI and Netconf
Stage3: Custom Yang modules and custom Yang translator
Stage4: Custom Yang config provisioning using CLI and Netconf

By @valjeanchan, @jnpr-raylam 

Closes https://github.com/nre-learning/antidote/issues/165

Mierdin commented:
Thanks @jnpr-raylam. Will begin a full review shortly. Thanks for working on this.

Can you reword the title, though? "Automating with Openconfig" doesn't really describe what's going on here, or what openconfig really is. "automating" can apply broadly to anything on this site. Maybe something like "Vendor-Neutral Network Configuration with OpenConfig" perhaps?

Mierdin commented on lessons/lesson-26/stage3/guide.md:
```diff
@@ -0,0 +1,79 @@
+
+In this lesson, we're going to define a custom YANG module to create new config stanza, and a custom YANG translation script to translate custom defined configuration to Junos config.
+
+#### Loading custom YANG modules
```
```suggestion
#### Loading Custom YANG Modules
```

Mierdin commented on lessons/lesson-26/stage4/guide.md:
```diff
@@ -0,0 +1,89 @@
+
+### Chapter 4 - Custom YANG config provisioning using CLI and NETCONF
+
+In this stage, we will configure the L3VPN services using the `vpn-services` config knob defined by our custom YANG model.
```
```suggestion
In this chapter, we will configure the L3VPN services using the `vpn-services` config knob defined by our custom YANG model.
```

Mierdin commented on lessons/lesson-26/stage4/guide.md:
```diff
@@ -0,0 +1,89 @@
+Same as before, press `ENTER` when you're asked to restart the `cli`
+
+
+#### Config using custom YANG
```
```suggestion
#### Configuration Using Custom YANG
```

Mierdin commented on lessons/lesson-26/stage4/guide.md:
````diff
@@ -0,0 +1,89 @@
+```
+<button type="button" class="btn btn-primary btn-sm" onclick="runSnippetInTab('vqfx', 2)">Run this snippet</button>
+
+#### Config verification
````
```suggestion
#### Config Verification
```

Mierdin commented on lessons/lesson-26/stage4/guide.md:
````diff
@@ -0,0 +1,89 @@
+```
+<button type="button" class="btn btn-primary btn-sm" onclick="runSnippetInTab('vqfx', 4)">Run this snippet</button>
+
+#### config custom YANG configration using NETCONF
````
```suggestion
#### Config Custom YANG Configuration using NETCONF
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,34 @@
+
+utilities:
+- name: linux
+  image: raylam/utility
```
This will be switched to `antidotelabs/utility` before prod

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,34 @@
+connections:
+- a: vqfx
+  b: linux
+  subnet: 10.1.0.0/24
```
```suggestion
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,34 @@
+
+devices:
+- name: vqfx
+  image: raylam/vqfx-full-17.4:190204
```
This will be switched to `antidotelabs/vqfx-full:<some tag>` before prod.

Mierdin commented:
@jnpr-raylam Ah, nevermind I obviously don't have push access to your repo.

You can either grant me push access and I'll commit the easy typo ones, or you can. Most of my comments are simple like that.

Thanks for this PR!

jnpr-raylam commented on lessons/lesson-26/stage5/configs/vqfx.txt:
```diff
@@ -0,0 +1,63 @@
+version 17.4R1.16;
```
stage 5 had been removed in latest commit, I guess you're viewing the old version?

Mierdin replied:
Yes, I believe this was one of my earliest comments in the review, and I had started it before you added that last commit.

jnpr-raylam commented on lessons/lesson-26/stage5/guide.md:
```diff
@@ -0,0 +1,7 @@
+## Automating with Openconfig
```
stage 5 had been removed in latest commit, I guess you're viewing the old version?

jnpr-raylam commented on lessons/lesson-26/stage1/guide.md:
````diff
@@ -0,0 +1,54 @@
+
+```request system software add junos-openconfig-XX.YY.ZZ.JJ-signed.tgz no-validate```
+
+**Note:** _Normally it takes less than a minute to complete the installation with physical or virtual Junos devices.  However, in NRE Labs, the vQFX run over container with limited virtualization capability, so it will take additional time to install._ **Therefore, we prepared an vQFX image with OpenConfig package pre-installed.**
````
This openconfig QFX image shares the same base with JET's one, but added the openconfig image.

For JET lesson, we have to use a new image because the original QFX image is lite-mode (with RE only), but JET must be run in vQFX full mode (RE+PFE).  So I packaged another container image, running two qemu-img commands to bring up RE + PFE.   For full mode, the interfaces name is changed from em2/em3/em4/.... to xe-0/0/[0-9], so for existing lesson, I think it's better to keep using existing lite-mode image due to smaller footprint, and also don't need to update the interface naming in all guide.md/scripts

In my environment, if I run the vQFX on top of KVM directly with hardware virtualization enabled, the openconfig installation time is very fast.  However, in NRE labs, the vQFX is running in VM over container over VM, without hardware virtualization. (I guess the google platform doesn't support nested virtualization), from my testing, it takes over 10 minutes to install. That's why I made another image with openconfig pre-installed.

Mierdin replied:
Yes this is why I like to talk things over first :smile:  - @mwiget has been hard at work creating a [new vqfx image with PFE](https://github.com/nre-learning/antidote/tree/master/images/container-vqfx) for exactly this use case. At this point, my recommendation would be to look at the work he contributed in #194 and #195, and see if there's anything from your existing image sources (which I would need you to contribute anyways if we went with your image) that you feel like would add value there. I have yet to finish testing on that image, but Marcel worked on that for some time so when we encounter use cases like this that require a full image, I'd rather have a single choice, so that we don't have a bunch of images floating around, confusing future contributors.

Next, we absolutely can and should be using hardware-accelerated virtualization. The original reason we built on top of GCP was so that we could enable this. So if this has been disabled somewhere along the chain (either at the VM layer or somewhere in the configuration of kubernetes or our image), we should fix it. This is definitely possible.

So, the takeaway is this. I am going to - in the next 24 hours - finish testing @mwiget new full vqfx image and ensure that it not only works functionally within NRE Labs but also has full hardware support - the intention from the beginning was to preserve this, so if this isn't working, I'll fix that. Once done, you'll need to reconcile the image work you've done with that of @mwiget (if applicable), and if you feel like any additional image changes need to be made, include them in this PR (no changes is fine too - I just don't see your source in this PR so I have no idea what you've done there).

Sound good?

jnpr-raylam commented on lessons/lesson-26/stage1/guide.md:
```diff
@@ -0,0 +1,54 @@
+## Vendor neutral configuration provisioning using YANG and OpenConfig
```
Agree with the change, and have updated all the lesson name in guide.md/syringe.yml file.

jnpr-raylam commented:
Thanks, I have updated the lesson according to your comments, and please review my comment regarding the vqfx image.
Also, I've added you as the collaborator of my repo.

Mierdin commented:
@jnpr-raylam Thanks. And everything except for the vqfx image conversation looks to be resolved, so stay tuned for now, I'll let you know when I have a new image for you to test with. Once you've done that, assuming everything is satisfactory, I will consider this ready to merge.

Mierdin commented:
A quick update - I'm having a discussion in https://github.com/nre-learning/syringe/pull/93 regarding one of things that's standing in the way of us using the new full vqfx image. Once that conversation is worked out I believe we'll be ready to use the new image. Stay tuned there for updates for now.

In the meantime, I'd like to keep this PR open so we can merge it when it's totally ready to go. I'd feel comfortable merging it now, but I'd have to keep it in PTR until the new image is ready to use, at which point we'd have to update this lesson to use it anyways. I'm still committed to getting this in the next release so we're really all just waiting for this image stuff to get hashed out and then all the pieces will fall into place. Thanks for your patience.

Mierdin commented:
FYI I'm going to keep this open for now, but will likely move forward with the next release soon, so we can get the Terraform and JET lessons published, provided the JET lesson can work with the new image.

We're running into some challenges with OpenConfig on the new image that we're working on in #208 in case you're interested and would like to chime in.

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,33 @@
+
+utilities:
+- name: linux
+  image: raylam/utility
```
```suggestion
  image: antidotelabs/utility
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,33 @@
+
+devices:
+- name: vqfx
+  image: raylam/vqfx-full-17.4:190204
```
```suggestion
  image: antidotelabs/vqfx-full:18.1R1.9
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,33 @@
+---
+lessonName: Vendor-Neutral Network Configuration with OpenConfig
+lessonId: 26
+category: configuration
```
```suggestion
category: tools
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,33 @@
+lessonId: 26
+category: tools
+tier: local
+juniperSpecific: true
```
```suggestion
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,33 @@
+lessonName: Vendor-Neutral Network Configuration with OpenConfig
+lessonId: 26
+category: tools
+tier: local
```
```suggestion
tier: prod
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,32 @@
+lessonId: 26
+category: tools
+tier: prod
+
```
```suggestion
description: "In this lesson, we'll explore OpenConfig - an industry effort to standardize network configuration and telemetry models"
```

Mierdin commented on lessons/lesson-26/syringe.yaml:
```diff
@@ -0,0 +1,32 @@
+lessonId: 26
+category: tools
+tier: prod
+description: "In this lesson, we'll explore OpenConfig - an industry effort to standardize network configuration and telemetry models"
```
```suggestion
description: "In this lesson, we'll explore OpenConfig - an industry effort to standardize network configuration and telemetry models"
slug: openconfig

```

Mierdin commented:
Good news! The new vqfx image has all the openconfig wrinkles ironed out, so I'm going to merge this and prep it for inclusion in the release. Thanks for the great work!
